Adding initial draft for timeseries pipeline documentation (#201)
timescale/timescaledb-toolkit: docs/timeseries_pipeline_elements.md

Timescale timeseries objects are just a convenient and efficient way of tracking a single value over time and are detailed a bit more [here](timeseries.md).  One of our primary goals with timeseries is that they should be easy and efficient to perform basic operations on, and that is where pipelines enter the picture.  At its simplest, a pipeline is just a timeseries connected to a [pipeline element](#timeseries-pipeline-elements) via the pipeline operator `|>`.  However, most pipeline operations output new timeseries, so it's possible to chain many pipeline elements together such that the output from one element become the input to the next.

### A note on operator associativity and grouping

Due to limitations in the PostgresQL parser, custom operators are required to be left associative.  The following pipeline will always result in `elementA` being applied to `timeseries` and then `elementB` being applied to the result.

```SQL ,ignore
SELECT timeseries |> elementA |> elementB;
```

However, it is possible to explicitly group elements using parentheses:

```SQL ,ignore
SELECT timeseries |> (elementA |> elementB);
```

> This will result in a pipeline object being created from elements A and B, which will then be applied to the timeseries.  While we don't presently take maximum advantage of this internally, these multiple element pipelines should enable optimizations moving forward.  Therefore, this second form should be preferred where possible.

Reviewer: "Therefore, this second form should be preferred where possible"

This feels like guidance that's just way too general. Are there drawbacks? Is there a time you wouldn't want to do that? IDK, just not sure where to go with this as a user because I'm sure at some point it won't always work as I want??

Anyway, just some feedback to consider.

## Usage Example <a id="timeseries-pipeline-example"></a>

For this example let start with a table of temperatures collected from different devices at different times.

```SQL ,non-transactional,ignore-output
SET TIME ZONE 'UTC';
CREATE TABLE test_data(time TIMESTAMPTZ, device INTEGER, temperature DOUBLE PRECISION);
```

In order to have some nominally interesting data to look at, let's populate this table with random data covering 30 days of readings over 10 devices.

```SQL ,non-transactional,ignore-output
SELECT setseed(0.456);
INSERT INTO test_data 
    SELECT
        '2020-01-01 00:00:00+00'::timestamptz + ((random() * 2592000)::int * '1 second'::interval),
        floor(random() * 10 + 1),
        50 + random() * 20
    FROM generate_series(1,10000);
```

> Now suppose we want to know how much the temperature fluctuates on a daily basis for each device.  It is actually simple to do this using timeseries and pipelines:

Reviewer: ```suggestion
Now suppose we want to know how much the temperature fluctuates on a daily basis for each device.  Using timeseries and pipelines can simplify the process of finding the answer:
```
(not fully happy with the suggestion, but we generally try to avoid saying things are "easy" and "simple". It's not to everyone. ;-)

```

## Current Pipeline Elements(A-Z) <a id="timeseries-pipeline-elements"></a>

As of the current timescale release, these elements are all [experimental](/docs/README.md#tag-notes).

> - [delta](#timeseries_pipeline_delta)
> - [fill_holes](#timeseries_pipeline_fill_holes)
> - [lttb](#timeseries_pipeline_lttb)
> - [resample_to_rate](#timeseries_pipeline_resample_to_rate)
> - [sort](#sort)

---

## **delta** <a id="timeseries_pipeline_delta"></a>
```SQL ,ignore
delta(
) RETURNS TimeseriesPipelineElement
```

This element will return a new timeseries where each point is the difference between the current and preceeding value in the input timeseries.  The new series will be one point shorter as it will not have a preceding value to return a delta for the first point.

> ### Required Arguments <a id="timeseries_pipeline_delta-arguments"></a>

Reviewer: Just FYI, when you move these over to our docs, we use the form of

`[](#timeseries_pipeline_delta-arguments)` for anchor links because we have a few different ways of creating the in-page links so this allows consistency.

Valid fill methods are:
| Method | Description |
|---|---|
| `locf` | Fill gaps with the last valid preceeding value. |
| `interpolate` | Compute the missing value linearly from the immediately bounding values |

### Required Arguments <a id="timeseries_pipeline_fill_holes-arguments"></a>
|Name| Type |Description|
|---|---|---|
| `fill_method` | `TEXT` | Case insensitive match for one of the fill methods above. |
<br>

### Pipeline Execution Returns <a id="timeseries_pipeline_fill_holes-returns"></a>

|Column|Type|Description|
|---|---|---|
| `timeseries` | `Timeseries` | This creates a complete normal timeseries (no missing values) from the input series.. |
<br>

### Sample Usage <a id="timeseries_pipeline_fill_holes-examples"></a>
```SQL
SELECT time, value
FROM toolkit_experimental.unnest_series(
    (SELECT toolkit_experimental.timeseries('2020-01-01'::timestamptz + step * step * '1 hour'::interval, step * step) 

>         |> (toolkit_experimental.resample_to_rate('nearest', '1 hour', true)

Reviewer: can this be addressed before adding it into an example?

        |>  toolkit_experimental.fill_holes('locf'))
    FROM generate_series(1, 3) step)
);
```
```output
          time          | value 
------------------------+-------
 2020-01-01 01:00:00+00 |     1
 2020-01-01 02:00:00+00 |     1
 2020-01-01 03:00:00+00 |     1
 2020-01-01 04:00:00+00 |     4
 2020-01-01 05:00:00+00 |     4
 2020-01-01 06:00:00+00 |     4
 2020-01-01 07:00:00+00 |     4
 2020-01-01 08:00:00+00 |     4
 2020-01-01 09:00:00+00 |     9
```

---

## **lttb** <a id="timeseries_pipeline_lttb"></a>
```SQL ,ignore
lttb(
    resolution int,
) RETURNS TimeseriesPipelineElement
```

> This element will return a [largest triangle three buckets](lttb.md#description) approximation of a given timeseries.  It's behavior is the same as the lttb function documented [here](lttb.md#lttb), save that it expects the series to be sorted.

Reviewer: ```suggestion
This element will return a [largest triangle three buckets](lttb.md#description) approximation of a given timeseries.  Its behavior is the same as the lttb function documented [here](lttb.md#lttb), save that it expects the series to be sorted.
```

<br>

### Pipeline Execution Returns <a id="timeseries_pipeline_resample_to_rate-returns"></a>

|Column|Type|Description|
|---|---|---|
| `timeseries` | `Timeseries` | A new pipeline with `interval` spaced points generated from the input series |
<br>

### Sample Usage <a id="timeseries_pipeline_resample_to_rate-examples"></a>
```SQL
SELECT time, value::numeric(4,2)
FROM toolkit_experimental.unnest_series(
    (SELECT toolkit_experimental.timeseries('2020-01-01'::TIMESTAMPTZ + step *step * step * '1 minute'::interval, step) 
        |> toolkit_experimental.resample_to_rate('weighted_average', '1 hour', true) 
    FROM generate_series(1,10) step)
);
```
```output
          time          | value 
------------------------+-------
 2020-01-01 00:00:00+00 |  1.59
 2020-01-01 01:00:00+00 |  4.00
 2020-01-01 02:00:00+00 |  5.00
 2020-01-01 04:00:00+00 |  6.00
 2020-01-01 06:00:00+00 |  7.00
 2020-01-01 09:00:00+00 |  8.00
 2020-01-01 12:00:00+00 |  9.00
 2020-01-01 17:00:00+00 | 10.00
```

---

## **sort** <a id="timeseries_pipeline_sort"></a>
```SQL ,ignore
sort(
) RETURNS TimeseriesPipelineElement
```

> This element takes in a timeseries and returns a timeseries consisting of the same points, but in order of increasing time values.

Reviewer: Always increasing? Ever a need for decreasing time? (mostly just curious)
